drivers/wdm/audio/backpln/portcls/irpstream.cpp

```diff
@@ -534,84 +570,115 @@ CIrpQueue::GetMappingWithTag(
     OUT PULONG  Flags)
 {
     PKSSTREAM_DATA StreamData;
+    KIRQL OldLevel;
+    ULONG Size;
+    LPBYTE Data;
 
     /* sanity checks */
     PC_ASSERT(PhysicalAddress);
     PC_ASSERT(VirtualAddress);
     PC_ASSERT(ByteCount);
     PC_ASSERT(Flags);
 
+
+    KeAcquireSpinLock(&m_IrpListLock, &OldLevel);
```

Comment:
Why does the spin lock for the IRP list need to be acquired until the very end of this function? Isn't it possible to use again these ExInterlocked* functions as before? (and doing the explicit KeAcquire/ReleaseSpinLock calls only in RemoveHeadList_IRP ?)

Reply:
The entire function needs to be atomic not just the individual variables. The existing lock-less code was definitely wrong. It was wrong in the first few lines, if two threads ran the code below at the same time two IRPs could end up being pulled from the queue and one of them would be lost since m_Irp is a shared variable.

if (!m_Irp) { m_Irp = KsRemoveIrpFromCancelableQueue(); }